File: apina-core/src/main/java/fi/evident/apina/utils/ResourceUtils.java
package fi.evident.apina.utils;

import java.io.*;
import java.nio.charset.Charset;

public final class ResourceUtils {

    private ResourceUtils() { }

    public static String readResourceAsString(String path, Charset charset) throws IOException {
        try (Reader reader = openResourceAsReader(path, charset)) {
            StringBuilder sb = new StringBuilder();
            char[] buffer = new char[1024];
            int n;

            while ((n = reader.read(buffer)) != -1)
                sb.append(buffer, 0, n);

            return sb.toString();
        }
    }

    public static Reader openResourceAsReader(String path, Charset charset) throws FileNotFoundException {
        return new InputStreamReader(openResourceAsStream(path), charset);
    }

    public static InputStream openResourceAsStream(String path) throws FileNotFoundException {
        ClassLoader classLoader = ResourceUtils.class.getClassLoader();
        InputStream in = classLoader.getResourceAsStream(path);
        if (in != null)
            return in;
        else
            throw new FileNotFoundException("could not find classpath resource: " + path);
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/utils/ResourceUtils.kt
package fi.evident.apina.utils

import java.io.FileNotFoundException
import java.io.InputStream
import java.nio.charset.Charset

fun readResourceAsString(path: String, charset: Charset = Charsets.UTF_8): String =
        openResourceAsStream(path).reader(charset).use { it.readText() }

fun openResourceAsStream(path: String): InputStream =
        object {}.javaClass.classLoader.getResourceAsStream(path)
                ?: throw FileNotFoundException("could not find classpath resource: $path")
